Fixed #26974, #28887, #28990, #29614 -- Added support for B-Tree, Hash and SP-Gist indexes; BrinIndex autosummarize parameter.

pope1ni opened the pull request:
Tickets [#26974](https://code.djangoproject.com/ticket/26974), [#28887](https://code.djangoproject.com/ticket/28887), [#28990](https://code.djangoproject.com/ticket/28990) & [#29614](https://code.djangoproject.com/ticket/29614).

This should round out the support for PostgreSQL indexes in Django.

There are a few small bits of refactoring to pave the way for the additional index types. I've also added some version checks to `create_sql()` for `BrinIndex` and `GinIndex`.

A few things for some consideration:

- [x] Can we somehow add support for `fillfactor` to `django.db.models.indexes.Index`?
  - _Perhaps we will be forced to add `django.contrib.postgres.indexes.BtreeIndex`..._
- [x] ~~Is there some way we can refactor out the option handling? It's not very DRY...~~
- [x] ~~Should there be better validation of the options? e.g. `fillfactor` is an integer between 10 and 100.~~
- [x] Tests still need adding for the feature flag checking.
  - ~~_Perhaps we should add a `check_supported()` method to `django.db.models.indexes.Index`?_~~

pope1ni commented:
@timgraham: Do any of the jenkins builders have support for PostgreSQL 10?

pope1ni commented:
@atombrella, @akki: Would you be willing to review this and discuss the things listed above given that you both worked on the existing postgres index support?

atombrella commented:
@pope1ni About the first remark about `fillfactor`, I think it's unlikely that this could be something that's in the base class. `CREATE INDEX` is not standardized, and the syntax for customization differs quite a lot among the different databases. For instance, SQLite doesn't support any customization.

pope1ni commented:
> I think it's unlikely that this could be something that's in the base class.

Yeah. I figured that would probably be the case. I suppose that we could create a `BTreeIndex` just for PostgreSQL to support custom options.

In which case the base `Index` class can be seen as providing the best default index for each backend.

timgraham commented:
Regarding the first commit "Renamed kwargs parameter to options for PostgreSQL index classes.", I don't think that's needed. It's consistent with the model field docs and code.

Regarding the version checks, I think me and akki discussed that having the database raise an exception is fine.

PostgreSQL 10 isn't on any CI servers at this time.

Are all these commits dependent on each other? It's easier to review and merge in small chunks when possible.

pope1ni commented:
> It's consistent with the model field docs and code.

Ok. I'll ditch that one.

> ...having the database raise an exception is fine.

```sql
example=# CREATE INDEX unknown_idx ON auth_user USING unknown (username);
ERROR:  access method "unknown" does not exist
example=# CREATE INDEX unknown_idx ON auth_user USING btree (username) WITH (unknown = off);
ERROR:  unrecognized parameter "unknown"
```

Hmm. This feels a little bit vague to me.

If we decide not to have the checks, I feel that we should at least mention what versions of PostgreSQL are required in the documentation for the various indexes/options?

I did find this [comment](https://github.com/django/django/pull/7950#discussion_r122726825) on not validating the values provided which I agree with. (So we can ignore my third list item on validating `fillfactor`, especially as the acceptable range of values sounds as though it is index type-specific: [documentation](https://www.postgresql.org/docs/current/static/sql-createindex.html#SQL-CREATEINDEX-STORAGE-PARAMETERS).)

> PostgreSQL 10 isn't on any CI servers at this time.

Ok. I will ensure that the tests pass for me locally then.

> Are all these commits dependent on each other?

Will re-jig things to address the other comments, so that they will be independent of each other (with the exception of the refactoring commits being a pre-requisite.)

akki commented on django/contrib/postgres/indexes.py:
```diff
@@ -1,47 +1,65 @@
+        # longer than the usual 3 character limit. The 30 character limit for
+        # cross-database compatibility isn't applicable to indexes only usable
+        # on PostgreSQL.
+        return Index.max_name_length - len(Index.suffix) + len(self.suffix)
```
Can we restrict the suffixes of these indexes to 3 characters as well (to be consistent with other non-postgres code)?
What benefits are we gaining from increasing the limit of the suffix part of the index name - and why _just_ the suffix part then, why not allow longer names as well if we're making it postgres-specific.

Although I would try to keep the rules as consistent as possible and simply give 3 chars suffixes to these indexes as well; Or maybe I am missing something here?

pope1ni replied:
I think not for a few reasons:

- The ship has sailed - we are already using longer suffixes for PostgreSQL, e.g. `brin` and `like`.
- The name truncating code is part of the main base `Index`. It would need separate logic for PostgreSQL which is not impossible, but beyond the scope of these tickets.
- Suffixes such as `brn`, `hsh`, `gst`, `spg`, `lik` would just be unclear for no good reason other than unrelated backends having a name length restriction.

timgraham replied:
A reason for the longer suffixes was to simplify inspectdb. Using different names would require mapping the names that PostgreSQL returns to whatever names we picked.

akki replied:
I see. Thanks @pope1ni and @timgraham

pope1ni commented:
Ok. So I have rearranged into the following parts:

- **Commits 1-3:** Minor refactoring to clean things up ahead of the new indexes.
- **Commits 4-6:** Added new indexes and options; depends on above commits, independent of each other except release notes.
- **Commits 7-9:** Added version checks for each index; some depend on new indexes and options above, commits 8 and 9 depend on commit 7 for new imports.

I hope that this makes things easier. I think that these are good to go.

The only thing missing now is `fillfactor` for `btree` indexes which are implemented via the generic `Index` class. Do people think that a `BTreeIndex` should be provided for this?

timgraham commented:
Maybe https://github.com/django/django/pull/9561 is a more DRY version of commit 2 and 3? Will it play nicely with the rest of your patches?

And for feature detection, perhaps we should add some helpers: https://github.com/django/django/pull/9562.

pope1ni commented:
Your suggested changes look good to me.

I'll update this PR after #9561 is merged. I'll have to handle the checks in commits 7-9 differently, but that will be trivial to change. I'll also add `is_postgresql_100` similar to the change in #9562 for commits 4 and 6.

pope1ni commented:
Hi Tim,

Did you reorder the commits as 5305b2c above depends on e1aba31 but comes before?

Is there anything else that needs to be done or do we think this is ready?

timgraham commented:
If you've amended some commits, GitHub reorders them somehow. Don't worry about that.

pope1ni commented:
Ok. Hadn't come across that before. Thanks.

timgraham commented on docs/ref/contrib/postgres/indexes.txt:
```diff
@@ -78,3 +78,24 @@ available from the ``django.contrib.postgres.indexes`` module.
+    .. versionadded:: 2.1
+
+    Creates a `hash index
+    <https://www.postgresql.org/docs/current/static/hash.html>`_.
```
This page doesn't exist and after a quick search, I didn't see a suitable replacement.

pope1ni replied:
I was a bit presumptuous here. Will have a look for an alternative link or remove.

pope1ni replied:
There only seem to be specific sections for brin, gin, gist and spgist: https://www.postgresql.org/docs/current/static/internals.html

timgraham commented on docs/ref/contrib/postgres/indexes.txt:
```diff
@@ -78,3 +78,24 @@ available from the ``django.contrib.postgres.indexes`` module.
+
+    .. note:
+
+        Although hash indexes have been available in PostgreSQL for a long
```
I'm unsure about this. As far as I see, the tests pass on older versions.

pope1ni replied:
Yes, they will. Hash indexes work on older versions of PostgreSQL, but they were unsafe, and I thought it would be irresponsible to allow people to use them unless they are using version 10 or later. If we decide that Django shouldn't police this, then we still need a big fat warning.

See the first caution at https://www.postgresql.org/docs/9.6/static/indexes-types.html

Also read this article: http://rhaas.blogspot.co.uk/2017/09/postgresqls-hash-indexes-are-now-cool.html

timgraham replied:
Oh, I didn't notice the subsequent commit in this PR that added the version check -- I would have put that in the same commit. I don't know if someone might want to use hash indexes on older versions of PostgreSQL but it doesn't really feel like Django's place to block usage of it.

pope1ni replied:
Sorry - the version checks I moved into subsequent commits so that they could be dropped if decided against. This was based on previous discussion with you above - I'm not sure we've had a firm decision on that yet since my most recent comments that the errors returned are too vague when index types or options do not exist.

We could allow hash indexes with <10 but they don't replicate, or backup properly IIRC, and if PostgreSQL crashes they are highly likely to become corrupt. This makes their use quite limited...

timgraham replied:
For Hash indexes, I think a documentation warning is sufficient. It seems unlikely that someone would choose an index without reading about how it works. I guess the other version checks are fine.

pope1ni replied:
Ok. We'll go with that then.

timgraham commented on tests/postgres_tests/test_indexes.py:
```diff
@@ -132,6 +134,46 @@ def test_deconstruction_no_customization(self):
+    def test_suffix(self):
+        self.assertEqual(HashIndex.suffix, 'hash')
+
+    def test_eq(self):
```
I'm unsure if we need eq/name autogeneration tests for every subclass when those methods aren't overridden. Thoughts?

pope1ni replied:
Agreed. Will look into tidying this up.

timgraham commented on tests/postgres_tests/test_indexes.py:
```diff
@@ -132,6 +134,46 @@ def test_deconstruction_no_customization(self):
+            }
+        )
+
+    def test_deconstruction_no_customization(self):
```
As above -- it seems somewhat useless to implement this test for every subclass. At the least, maybe there could be a base test class that implements some common tests if that's warranted.

pope1ni replied:
Agreed. Will address this and the previous together.

timgraham commented:
Nick, did you forget about this or just lacking the time?

pope1ni commented:
Tim. Was just lacking the time, but should get a chance over the next couple of weeks to come back to this. I think this was mostly there - mainly needed some dedupe in the tests which I've started.

pope1ni commented:
Ok. So I have updated this and addressed the following points:
- [x] Rebased and updated release notes and version annotations to 2.2
- [x] Removed the version check for HashIndex and updated the warning in the documentation.
- [x] Removed the unnecessary tests where `Index.__eq__()` is not overridden.
- [x] Refactored basic tests for `deconstruct()` into a mixin to make things more DRY.
- [x] Refactored basic tests for name auto-generation into the same mixin.
- [x] Added a new `BTreeIndex` class to enable setting `fillfactor` parameter.

I've pushed everything up and am waiting to see what comes out of CI.

I know that at least one test is currently failing for `BTreeIndex` due to how introspection works. Any ideas on how to tackle that would be well received.

carltongibson commented:
Well done 👍🏽 Awesome effort. I will have another look tomorrow.

pope1ni commented:
> Frankly, this is super.

Thanks!

> I felt a bit sorry for poor Hash Index not getting a link to the pg docs

Yeah, there isn't really a separate page for `btree` either. Nor a sensible fragment in any of the other more generic pages. I think this is because `hash` and `btree` are well-known concepts in many different databases whereas the others are specific to PostgreSQL.

timgraham commented:
Merged in fa54ebc722a64d5b1b2f76301a15bd4e1a28b3ee, d526b07784d8caa208055c1f2bd1cedc88fb52dc, 4c36e9e492b5c127d839ebf280fdf5556af8e824, 4c36e9e492b5c127d839ebf280fdf5556af8e824, 4c36e9e492b5c127d839ebf280fdf5556af8e824, 4c36e9e492b5c127d839ebf280fdf5556af8e824, and 743d28f5539b17d6e39bd37c6e3df5628b470cac. Thanks!
